### PR TITLE
Add architecture-rules.md references to Test-Writer, Issue-Planner, and Solution-Designer

### LINKED ISSUE #334: Add architecture-rules.md references to Test-Writer, Issue-Planner, and Solution-Designer agents
## Summary

`architecture-rules.md` is a key consumer-repo artifact that defines layer boundaries, dependency rules, and naming conventions. An audit of all 14 agents shows that 8 already reference `.github/architecture-rules.md`, but **Test-Writer**, **Issue-Planner**, and **Solution-Designer** — three agents whose outputs directly depend on architectural awareness — do not.

### Agents that reference `.github/architecture-rules.md` (8 of 14)

| Agent | References | Usage |
|---|---|---|
| Code-Conductor | 3 | Docs-misrepresentation check, Tier 2 structural validation, routing to Refactor-Specialist |
| Code-Critic | 3 | Architecture compliance check, prosecution checklist, context file list |
| Code-Smith | 2 | Pure domain layer rule, extraction & extension principles |
| Refactor-Specialist | 3 | Dependency injection, layer placement verification, context file list |
| Research-Agent | 2 | Understanding boundaries, critical context file |
| Specification | 1 | Architecture and layer placement |
| UI-Iterator | 1 | Context file list |
| Doc-Keeper | 1 | Project guidance loading |

### Agents with no `.github/architecture-rules.md` reference (6 of 14)

| Agent | Gap severity | Rationale |
|---|---|---|
| **Test-Writer** | **High** | Tests should respect layer boundaries for mocking strategy (mock at layer boundary, not across layers), integration test scoping, and test file organization |
| **Issue-Planner** | **Medium** | Plan decomposition benefits from layer awareness — steps scoped to a single layer are cleaner and more testable |
| **Solution-Designer** | **Medium** | Proposes architecture options — should consult existing project architecture constraints for compatible designs |
| Experience-Owner | Low | Customer-focused; architecture is implementation detail |
| Process-Review | Low | Meta-analysis of workflow execution, not code |
| Code-Review-Response | Low | Judge role; prosecution findings already cite architecture rules |

Note: Solution-Designer references `examples/{stack}/architecture-rules.md` (Hub/Consumer gate, line 108) but NOT the consumer's own `.github/architecture-rules.md` for design exploration.

### Test-Writer — Quality Gates integration

**Insertion point**: Quality Gates section (around line 85), plus Skills Reference section (around line 284).

**Why Quality Gates, not Integration Test Rule**: The Integration Test Rule section has a clear anti-mock-helper mandate ("Integration tests MUST call actual production code paths, not mock helpers"). Inserting "mock at layer boundaries" guidance there creates interpretive tension — the two concepts are compatible (boundary mocking ≠ mock helpers) but their juxtaposition risks muddying the existing rule. Quality Gates already lists testability standards.

**Add to Quality Gates** (~3 lines):
- Test file organization should mirror architecture layers when the project defines them (see `.github/architecture-rules.md`)
- Mock/stub at layer boundaries (interfaces between layers), not deep internals — this keeps tests resilient to refactoring within a layer
- Integration test scope should exercise adjacent layers together per architecture-rules dependency direction

**Add to Skills Reference** (~1 line):
- `.github/architecture-rules.md` — Layer boundaries, dependency rules, and naming conventions

**Format**: Use dash-list style in Skills Reference (consistent with UI-Iterator, Refactor-Specialist, Research-Agent patterns). Use bullet style in Quality Gates (consistent with existing bullets there).

### Issue-Planner — Plan style guide integration

**Insertion point**: `<plan_style_guide>` only (NOT `<research_instructions>`).

**Why not research_instructions**: Research-Agent (the subagent dispatched by Issue-Planner's discovery phase) already marks `.github/architecture-rules.md` as **CRITICAL** in its own References section ([Research-Agent.agent.md line 203]). Adding it to `<research_instructions>` duplicates the directive to the same subagent, increasing maintenance surface for no behavioral gain.

**Add to plan_style_guide** (~2 lines, as a new bullet):
- When a step crosses a layer boundary (as defined in `.github/architecture-rules.md`), note the dependency direction and verify it aligns with the project's documented architecture rules. Scope steps to a single layer where feasible.

**Wording rationale**: Avoids soft "where practical" qualifier. Uses an actionable trigger ("when a step crosses a layer boundary") and a concrete action ("note the dependency direction and verify it aligns").

### Solution-Designer — Collaboration Pattern integration

**Insertion point**: Collaboration Pattern > Research step (step 1, around line 120), NOT "Load Skills First".

**Why not Load Skills First**: That section's semantic contract is skills and instruction files. `.github/architecture-rules.md` is project context, not a skill. The Collaboration Pattern > Research step already directs searching design docs and patterns — architecture rules fit naturally there.

**Add to Research step** (~1 line):
- Check `.github/architecture-rules.md` for existing architecture constraints, layer boundaries, and dependency rules

### Rejected Alternatives

1. **Integration Test Rule insertion for Test-Writer**: Rejected — creates interpretive tension with the adjacent anti-mock-helper mandate.
2. **`<research_instructions>` addition for Issue-Planner**: Rejected — redundant with Research-Agent's own CRITICAL reference to the same file.
3. **"Load Skills First" insertion for Solution-Designer**: Rejected — section semantic contract is skills/instructions, not general project context.
4. **New dedicated sections in any agent**: Rejected — over-engineering for a gap-fill that fits naturally into existing sections.
5. **Defensive logic for missing file**: Rejected — precedent-consistent with all 8 existing agents; file is part of standard /setup flow.
6. **Including Experience-Owner, Process-Review, Code-Review-Response**: Rejected — low-gap-severity agents whose outputs don't produce architecture-sensitive artifacts.

## Acceptance Criteria

- [ ] Test-Writer Quality Gates section includes layer-aware test guidance (organization, mocking at boundaries, integration scope)
- [ ] Test-Writer Skills Reference section lists `.github/architecture-rules.md`
- [ ] Issue-Planner `<plan_style_guide>` includes layer-boundary step scoping with actionable trigger
- [ ] Solution-Designer Collaboration Pattern > Research step includes architecture-rules context loading
- [ ] Reference format is consistent with local context patterns in each agent (dash-list in reference sections, bullets in Quality Gates, sentence form in prose)
- [ ] Quick-validate passes after changes
- [ ] No existing agent references are broken or duplicated

## Testing Scope

Documentation-only change (agent `.md` files). No unit/integration/E2E tests applicable. Validation: `quick-validate.ps1` structural checks only.

## Adversarial Design Challenge Summary

3-pass prosecution (2 design perspectives, 1 product-alignment). 9 findings total.

**Incorporated (6)**:
- F1: Mock guidance moved from Integration Test Rule → Quality Gates to avoid interpretive tension
- F2: Agent count corrected from 9 → 8
- F3: AC format loosened from single parenthetical to "consistent with local context patterns"
- F4: Issue-Planner wording strengthened from "where practical" to actionable trigger
- F5: Issue-Planner `<research_instructions>` dropped (redundant with Research-Agent)
- F7: Solution-Designer moved from "Load Skills First" to "Collaboration Pattern > Research"

**Dismissed (3)**:
- F6: Solution-Designer dual-reference (hub examples/ vs consumer .github/) — different files, different purposes, different sections; Hub/Consumer gate handles distinction
- F8: Expanding optional-file dependency blast radius (8→11) — precedent-consistent, standard /setup flow
- F9: Process-Review/Code-Review-Response omission — justified in gap analysis (low severity)

## Customer Experience

- **Customer surface**: None (internal agent configuration)
- **CE Gate**: Not applicable

### PR DESCRIPTION
**Pull request by Grimblaz**
## Summary

Adds `.github/architecture-rules.md` references to 3 agents that previously lacked them, improving architecture awareness during test writing, plan creation, and design exploration.

### Changed files

| File | Change |
|---|---|
| `.github/agents/Test-Writer.agent.md` | 3 architecture-aware bullets in Quality Gates + Skills Reference entry |
| `.github/agents/Issue-Planner.agent.md` | Layer-boundary scoping rule in `<plan_style_guide>` |
| `.github/agents/Solution-Designer.agent.md` | Architecture-rules check in Collaboration Pattern Research step |

### Validation evidence

- `markdownlint-cli2 --fix "**/*.md"` — pre-existing warnings only, no new issues
- `pwsh -NoProfile -NonInteractive -File .github/scripts/quick-validate.ps1` — 9/9 passed
- `Invoke-Pester .github/scripts/Tests/ -Output Minimal` — 385 passed, 0 failed, 1 skipped
- Scope check: 3 files changed, all planned (no unrelated files)

## Adversarial Review Scores

| Stage | Prosecutor | Defense | Judge rulings |
|---|---|---|---|
| Code Review | 2 pts (2 sustained) | 4 pts (4 disproved, 0 rejected) | 6 rulings |
| CE Review | ⏭️ N/A | ⏭️ N/A | ⏭️ N/A |
| Post-fix Review | ⏭️ N/A | ⏭️ N/A | ⏭️ N/A |

### Express-lane findings

- F4 (pass 2): Implicit file reference on Test-Writer line 98 → made explicit. `finding-sustained`.

## Prosecution Depth Summary

| Category | Depth | Rationale |
|---|---|---|
| architecture | full | — |
| security | full | — |
| performance | full | — |
| pattern | full | — |
| implementation-clarity | full | — |
| script-automation | full | — |
| documentation-audit | full | — |

Re-activated categories: none

## CE Gate

⏭️ CE Gate not applicable — internal agent configuration, no customer-facing surface

Process-Review: not invoked (no CE defects found)

## Pipeline Metrics

Closes #334

## Summary by Sourcery

Update agent configurations to reference shared architecture rules and enforce layer-aware behavior across planning, testing, and solution design workflows.

Enhancements:
- Add architecture-aware testing guidance to Test-Writer, including layer-aligned file organization and boundary-focused mocking.
- Require Issue-Planner plans to explicitly account for architecture layer boundaries and dependency direction when steps cross layers.
- Have Solution-Designer consult the central architecture-rules document when researching and framing design decisions.

### CODE REVIEW

File: .github/agents/Issue-Planner.agent.md [hunk 1]
```diff
@@ -275,6 +275,7 @@ Rules:
 - NO code blocks — describe changes, link to files/symbols
 - NO questions at the end — ask during workflow via #tool:vscode/askQuestions
 - Include execution metadata in plan steps (mode + requirement contract expectations) so implementers can execute without re-deriving process rules.
+- When a step crosses a layer boundary (as defined in `.github/architecture-rules.md`), note the dependency direction and verify it aligns with the project's documented architecture rules. Scope steps to a single layer where feasible.
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/agents/Issue-Planner.agent.md:278]: **Disposition: ❌ REJECT** (defense sustained, confidence: high)

The conditional trigger *"When a step crosses a layer boundary"* only fires when the consumer's `architecture-rules.md` defines layers — which all three template repos do. Repos without layers never trigger this rule. The issue's AC #3 explicitly specifies "layer-boundary step scoping with actionable trigger" — current wording matches.

File: .github/agents/Test-Writer.agent.md [hunk 2]
```diff
@@ -93,6 +93,9 @@ For PBT rollout policy and guardrails, follow `.github/skills/property-based-tes
 - Parameterized tests for formulas
 - Integration tests over unit tests where appropriate
 - PBT complements unit tests; does not replace requirement-focused examples
+- Test file organization should mirror architecture layers when the project defines them (see `.github/architecture-rules.md`)
+- Mock/stub at layer boundaries (interfaces between layers), not deep internals — keeps tests resilient to refactoring within a layer
+- Integration test scope should exercise adjacent layers together per the dependency direction in `.github/architecture-rules.md`
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/agents/Test-Writer.agent.md:98]: **Disposition: ❌ REJECT** (defense sustained, confidence: high)

These agents are hub agents that execute in **consumer repos**, not in the hub repo itself. The consumer's `.github/architecture-rules.md` — not the hub's — is the runtime target. All three consumer-repo templates (`examples/nodejs-typescript/`, `examples/python/`, `examples/spring-boot-microservice/`) define `## Layer Architecture` with explicit layer definitions, dependency direction, and boundary diagrams.

Line 96 includes the conditional guard *"when the project defines them"*, making it opt-in. In repos without layers, the guidance is inert. The terminology is accurate for the intended execution context.

The issue's acceptance criteria (AC #1) explicitly specify "layer-aware test guidance (organization, mocking at boundaries, integration scope)" — the current wording matches the AC.

File: .github/agents/Test-Writer.agent.md [hunk 3]
```diff
@@ -301,6 +304,10 @@ Generated stubs are pending by default (e.g., `return 'pending'` in cucumber.js)
 - Reference `.github/skills/verification-before-completion/SKILL.md`
 - Evidence before claims: run the repository's configured coverage command from `.github/copilot-instructions.md` before claiming coverage is sufficient
 
+**When checking architecture compliance:**
+
+- Consult `.github/architecture-rules.md` — Layer boundaries, dependency rules, and naming conventions
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/agents/Test-Writer.agent.md:309]: **Disposition: ❌ REJECT** (defense sustained, confidence: high)

Same consumer-repo context as the comment above — this is a navigational pointer listing what to look for when checking architecture compliance. In consumer repos, all three items (layer boundaries, dependency rules, naming conventions) are present in `architecture-rules.md`. In the hub, 2 of 3 match directly; "layer boundaries" is harmlessly absent — the agent reads what's there.

File: .github/agents/Solution-Designer.agent.md [hunk 4]
```diff
@@ -116,6 +116,7 @@ If the gate fires, redirect the proposal to the appropriate consumer artifact an
 For each design decision:
 
 1. **Research**: Search skills, `Documents/Research/`, `Documents/Design/`, `Documents/Decisions/`, and external patterns
+   - Check `.github/architecture-rules.md` for existing architecture constraints, layer boundaries, and dependency rules
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/agents/Solution-Designer.agent.md:119]: **Disposition: ❌ REJECT** (defense sustained, confidence: high)

This is a research checklist — tells the agent what to look for when consulting the file. In consumer repos, all three items (architecture constraints, layer boundaries, dependency rules) are defined. If any item is absent, the agent simply finds nothing for it and moves on — no misleading behavior.

**Comment by Grimblaz** [comment]
## GitHub Review Intake — Proxy Prosecution Summary

**Pipeline**: Proxy prosecution → Defense → Judge (Code-Review-Response)

### Finding Ledger

| ID | Source | File | Severity | Ruling | Confidence |
|----|--------|------|----------|--------|------------|
| G1 | sourcery-ai[bot] | (cross-file) | low (1 pt) | ❌ Defense sustained | high |
| G2 | copilot-pull-request-reviewer[bot] | Test-Writer.agent.md:96-98 | medium (5 pts) | ❌ Defense sustained | high |
| G3 | copilot-pull-request-reviewer[bot] | Test-Writer.agent.md:309 | medium (5 pts) | ❌ Defense sustained | high |
| G4 | copilot-pull-request-reviewer[bot] | Issue-Planner.agent.md:278 | medium (5 pts) | ❌ Defense sustained | high |
| G5 | copilot-pull-request-reviewer[bot] | Solution-Designer.agent.md:119 | medium (5 pts) | ❌ Defense sustained | high |

**Prosecutor**: 0 pts (0 sustained) | **Defense**: 21 pts (5 disproved) | **Judge rulings**: 5

### Defense Thesis

All five findings rest on a single analytical error: evaluating hub-agent guidance against the **hub's** `.github/architecture-rules.md` instead of the **consumer-repo** `architecture-rules.md` where these agents actually execute. All three consumer-repo templates (`examples/nodejs-typescript/`, `examples/python/`, `examples/spring-boot-microservice/`) define `## Layer Architecture` with layer definitions, dependency direction, and boundary diagrams — exactly the concepts referenced.

Additionally, the agent text uses conditional framing ("when the project defines them", "When a step crosses a layer boundary") ensuring graceful degradation in repos without layers.

### AC Cross-Check

The issue's acceptance criteria explicitly specify "layer-aware test guidance" (AC #1) and "layer-boundary step scoping with actionable trigger" (AC #3). Adopting the reviewers' suggested rewording would violate the AC.

**Result**: No fixes needed. PR remains merge-ready.
